File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.java
import org.w3c.dom.Node;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.sql.Array;
import java.util.*;



public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    static int N;
    static int [] arr;
    static ArrayList<Integer> list = new ArrayList<>();

    static boolean [] visited;

    static int max_value = (int) -1e9;


    static void dfs(int depth){

        if(depth == N){
            int sum = 0;
            for(int i=0; i<N-1; i++){
                sum += Math.abs(list.get(i) - list.get(i+1));
            }
            max_value =Math.max(max_value,sum);
            return;
        }

        for(int i=0; i<N; i++){
            if(visited[i]){
                continue;
            }
            visited[i] = true;
            list.add(arr[i]);
            dfs(depth+1);
            visited[i] = false;
            list.remove(list.size()-1);
        }


    }


    public static void main(String[] args) throws IOException {

        N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());

        arr = new int[N];
        visited = new boolean[N];
        for(int i=0; i<N; i++){
            arr[i] = Integer.parseInt(st.nextToken());
            visited[i] = false;
        }

        dfs(0);
        System.out.println(max_value);




    }
}

File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.py
def dfs(depth):
    if depth == n:
        current = 0
        for i in range(n - 1):
            current += abs(stack[i] - stack[i + 1])
        global max_value
        max_value = max(max_value, current)
        return

    for i in range(n):
        if not visited[i]:
            visited[i] = True
            stack.append(ls[i])
            dfs(depth + 1)
            stack.pop()
            visited[i] = False


n = int(input())
ls = list(map(int, input().split()))
visited = [False] * n
stack = []
max_value = -1
dfs(0)
print(max_value)